lld/test/ELF/aarch64-reloc-pauth.s

```diff
@@ -0,0 +1,228 @@
+##                              ^^^^ Discr = 46
+##                                    ^^ Key (bits 5..6) = DA
+##                                       ^^^^^^^^ No implicit val (rela reloc due to val wider than 32 bits)
+##                                                ^^^^ Discr = 47
+##                                                      ^^ Key (bits 5..6) = DA
+# HEX-NEXT: 0x00030470 00000000 30000020 00000000 31000020
+##                     ^^^^^^^^ No implicit val (rela reloc due val wider than 32 bits)
+##                              ^^^^ Discr = 48
+##                                    ^^ Key (bits 5..6) = DA
+##                                       ^^^^^^^^ No implicit val (rela reloc due to a preemptible symbol)
+##                                                ^^^^ Discr = 49
+##                                                      ^^ Key (bits 5..6) = DA
+# HEX-NEXT: 0x00030480 00000000 32000000 77000000 00330000
+##                     ^^^^^^^^ No implicit val (rela reloc due to a preemptible symbol)
+##                              ^^^^ Discr = 50
+##                                    ^^ Key (bits 5..6) = IA
+##                                         ^^^^^^ ^^ No implicit val (rela reloc due to odd offset)
+##                                                  ^^^^ Discr = 51
+# HEX-NEXT: 0x00030490 20774504 03003400 0020{{\ }}
+##                     ^^ Key (bits 5..6) = DA
+##                         ^^^^ ^^^^ Implicit val = 0x30445 = 0x30440 + 5 = .test + 5
+##                                  ^^^^ Discr = 52
+##                                         ^^ Key (bits 5..6) = DA
+
+#--- main.s
+
+.section .test, "aw"
+.p2align 3
+.quad (__ehdr_start + 1)@AUTH(da,42)
+.quad (.test + 2)@AUTH(ia,43)
+.quad (__ehdr_start - 3)@AUTH(ia,44,addr)
+.quad (__ehdr_start + 0x12345678)@AUTH(da,45)
+## Addend wider than 32 bits, not enough room for storing implicitly, would go to rela
+.quad (__ehdr_start + 0x123456789A)@AUTH(da,46)
+## Negative addend wider than 32 bits, not enough room for storing implicitly, would go to rela
+.quad (__ehdr_start - 0x123456789A)@AUTH(da,47)
+## INT32_MAX plus non-zero .test is wider than 32 bits, not enough room for storing implicitly, would go to rela
+.quad (.test + 0x7FFFFFFF)@AUTH(da,48)
+.quad (zed2 + 0x1111)@AUTH(da,49)
+.quad bar2@AUTH(ia,50)
+.byte 0x77
+.quad (__ehdr_start + 4)@AUTH(da,51)
+.byte 0x77
+.quad (.test + 5)@AUTH(da,52)
+
+#--- empty-relr.s
+
+## During relocation scanning, we might put some AUTH relocs into the packed
+## relocation section when not knowing the final implicit value to be stored in
+## the place to be relocated. If the value turns out to be wider than 32 bits
+## after computing the final layout, we convert the relocation from relr to
+## rela. A side effect is having an empty .relr.auth.dyn section if all the
+## packed relocations were converted to rela.
+## TODO: avoid empty .relr.auth.dyn in the output binary
```

Comment:
This TODO may be impractical to implement. We have such cases in tests, and we just avoid using "TODO". Simplify the comment:

.relr.auth.dyn relocations that do not fit 32-bit are moved to .relr.dyn. In this case .relr.auth.dyn will be makde empty, but removeUnusedSyntheticSections fails to remove the section.

Reply:
Fixed, thanks, see d32b8e32fc1c938e5a005997b832d215a308037d